firefox: Enable userChrome in about:config

MrQubo opened the pull request:
### Description

Enable "toolkit.legacyUserProfileCustomizations.stylesheets" in about:config if userChrome is not empty.
Resolves #5666

Change is not backwards compatible. It will break configs where `userChrome` is set to non-empty string and `settings."toolkit.legacyUserProfileCustomizations.stylesheets"` is set to false. It seems unlikely but I've added assertion nevertheless.

### Checklist

- [ ] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

#### Maintainer CC

@rycee @kira-bruneau

MrQubo commented:
Here's a more friendly diff without indentation changes:
```
diff --git a/modules/programs/firefox.nix b/modules/programs/firefox.nix
index 3bee6e0b..f946f1d2 100644
--- a/modules/programs/firefox.nix
+++ b/modules/programs/firefox.nix
@@ -301,7 +301,9 @@ in {
       };
 
       profiles = mkOption {
-        type = types.attrsOf (types.submodule ({ config, name, ... }: {
+        type = types.attrsOf (types.submodule ({ config, name, ... }:
+          let profile = config;
+          in {
           options = {
             name = mkOption {
               type = types.str;
@@ -704,7 +706,12 @@ in {
             };
 
           };
-        }));
+
+            config = {
+              settings."toolkit.legacyUserProfileCustomizations.stylesheets" =
+                mkIf (profile.userChrome != "") true;
+            };
+          }));
         default = { };
         description = "Attribute set of Firefox profiles.";
       };
@@ -757,6 +764,19 @@ in {
       }
 
       (mkNoDuplicateAssertion cfg.profiles "profile")
+
+      (let
+        assertProfile = profile:
+          profile.userChrome != ""
+          -> (profile.settings."toolkit.legacyUserProfileCustomizations.stylesheets"
+            != false);
+      in {
+        assertion = any assertProfile (attrValues cfg.profiles);
+        message = ''
+          'userChrome' won't work with
+          'settings."toolkit.legacyUserProfileCustomizations.stylesheets"' set to false.
+        '';
+      })
     ] ++ (mapAttrsToList
       (_: profile: mkNoDuplicateAssertion profile.containers "container")
       cfg.profiles);
diff --git a/tests/modules/programs/firefox/profile-settings-expected-userChrome-user.js b/tests/modules/programs/firefox/profile-settings-expected-userChrome-user.js
new file mode 100644
index 00000000..b0f8619c
--- /dev/null
+++ b/tests/modules/programs/firefox/profile-settings-expected-userChrome-user.js
@@ -0,0 +1 @@
+user_pref("toolkit.legacyUserProfileCustomizations.stylesheets", true);
diff --git a/tests/modules/programs/firefox/profile-settings-expected-userChrome-userChrome.css b/tests/modules/programs/firefox/profile-settings-expected-userChrome-userChrome.css
new file mode 100644
index 00000000..a2819786
--- /dev/null
+++ b/tests/modules/programs/firefox/profile-settings-expected-userChrome-userChrome.css
@@ -0,0 +1 @@
+#example { default: none; }
diff --git a/tests/modules/programs/firefox/profile-settings.nix b/tests/modules/programs/firefox/profile-settings.nix
index 8b781552..ceac20a9 100644
--- a/tests/modules/programs/firefox/profile-settings.nix
+++ b/tests/modules/programs/firefox/profile-settings.nix
@@ -153,6 +153,11 @@
           };
         };
       };
+
+      profiles.userChrome = {
+        id = 6;
+        userChrome = builtins.readFile ./profile-settings-expected-userChrome-userChrome.css;
+      };
     };
 
     nmt.script = ''
@@ -203,6 +208,14 @@
       assertFirefoxSearchContent \
         home-files/.mozilla/firefox/searchWithoutDefault/search.json.mozlz4 \
         ${./profile-settings-expected-search-without-default.json}
+
+      assertFileContent \
+        home-files/.mozilla/firefox/userChrome/user.js \
+        ${./profile-settings-expected-userChrome-user.js}
+
+      assertFileContent \
+        home-files/.mozilla/firefox/userChrome/chrome/userChrome.css \
+        ${./profile-settings-expected-userChrome-userChrome.css}
     '';
   };
 }
```

MrQubo commented:
Here's a more friendly diff without indentation changes:
<details>

```diff
diff --git a/modules/programs/firefox.nix b/modules/programs/firefox.nix
index 3bee6e0b..f946f1d2 100644
--- a/modules/programs/firefox.nix
+++ b/modules/programs/firefox.nix
@@ -301,7 +301,9 @@ in {
       };
 
       profiles = mkOption {
-        type = types.attrsOf (types.submodule ({ config, name, ... }: {
+        type = types.attrsOf (types.submodule ({ config, name, ... }:
+          let profile = config;
+          in {
           options = {
             name = mkOption {
               type = types.str;
@@ -704,7 +706,12 @@ in {
             };
 
           };
-        }));
+
+            config = {
+              settings."toolkit.legacyUserProfileCustomizations.stylesheets" =
+                mkIf (profile.userChrome != "") true;
+            };
+          }));
         default = { };
         description = "Attribute set of Firefox profiles.";
       };
@@ -757,6 +764,19 @@ in {
       }
 
       (mkNoDuplicateAssertion cfg.profiles "profile")
+
+      (let
+        assertProfile = profile:
+          profile.userChrome != ""
+          -> (profile.settings."toolkit.legacyUserProfileCustomizations.stylesheets"
+            != false);
+      in {
+        assertion = any assertProfile (attrValues cfg.profiles);
+        message = ''
+          'userChrome' won't work with
+          'settings."toolkit.legacyUserProfileCustomizations.stylesheets"' set to false.
+        '';
+      })
     ] ++ (mapAttrsToList
       (_: profile: mkNoDuplicateAssertion profile.containers "container")
       cfg.profiles);
diff --git a/tests/modules/programs/firefox/profile-settings-expected-userChrome-user.js b/tests/modules/programs/firefox/profile-settings-expected-userChrome-user.js
new file mode 100644
index 00000000..b0f8619c
--- /dev/null
+++ b/tests/modules/programs/firefox/profile-settings-expected-userChrome-user.js
@@ -0,0 +1 @@
+user_pref("toolkit.legacyUserProfileCustomizations.stylesheets", true);
diff --git a/tests/modules/programs/firefox/profile-settings-expected-userChrome-userChrome.css b/tests/modules/programs/firefox/profile-settings-expected-userChrome-userChrome.css
new file mode 100644
index 00000000..a2819786
--- /dev/null
+++ b/tests/modules/programs/firefox/profile-settings-expected-userChrome-userChrome.css
@@ -0,0 +1 @@
+#example { default: none; }
diff --git a/tests/modules/programs/firefox/profile-settings.nix b/tests/modules/programs/firefox/profile-settings.nix
index 8b781552..ceac20a9 100644
--- a/tests/modules/programs/firefox/profile-settings.nix
+++ b/tests/modules/programs/firefox/profile-settings.nix
@@ -153,6 +153,11 @@
           };
         };
       };
+
+      profiles.userChrome = {
+        id = 6;
+        userChrome = builtins.readFile ./profile-settings-expected-userChrome-userChrome.css;
+      };
     };
 
     nmt.script = ''
@@ -203,6 +208,14 @@
       assertFirefoxSearchContent \
         home-files/.mozilla/firefox/searchWithoutDefault/search.json.mozlz4 \
         ${./profile-settings-expected-search-without-default.json}
+
+      assertFileContent \
+        home-files/.mozilla/firefox/userChrome/user.js \
+        ${./profile-settings-expected-userChrome-user.js}
+
+      assertFileContent \
+        home-files/.mozilla/firefox/userChrome/chrome/userChrome.css \
+        ${./profile-settings-expected-userChrome-userChrome.css}
     '';
   };
 }
```
</details>

MrQubo commented:
I've tried running tests with `nix-shell --pure tests -A run.all` but I'm getting error `error: executing shell '/nix/store/p5441mp6wwpdh65qamaixvd092cwr45h-bash-interactive-5.2p26/bin/bash': Argument list too long`. :/

kira-bruneau commented on modules/programs/firefox.nix:
```diff
@@ -301,410 +301,418 @@ in {
 
-          };
-        }));
+            config = {
+              settings."toolkit.legacyUserProfileCustomizations.stylesheets" =
+                mkIf (profile.userChrome != "") true;
+            };
+          }));
```
Thanksl! :blush: 

I also just verified that `userContent` also needs this option, so I think it should be included here.

MrQubo replied:
It's the same options?? o.0 Cause it says "stylesheets" there.

MrQubo replied:
Ah, nvm, I was thinking that userContent refers to `userContent.js`.

IldenH commented on tests/modules/programs/firefox/profile-settings.nix:
```diff
@@ -1,6 +1,8 @@
 { config, lib, pkgs, ... }:
 
-{
+let userChromeContent = "#example { display: none; }";
+
+in {
```
```suggestion
let 

  userChromeContent = "#example { display: none; }";

in {
```

IldenH commented on modules/programs/firefox.nix:
```diff
@@ -301,410 +301,418 @@ in {
-        }));
+            config = {
+              settings."toolkit.legacyUserProfileCustomizations.stylesheets" =
+                mkIf (profile.userChrome != "") true;
```
```suggestion
                mkIf (profile.userChrome != "" || profile.userContent != "") true;
```

IldenH commented on modules/programs/firefox.nix:
```diff
@@ -301,410 +301,418 @@ in {
-              description = "Profile name.";
-            };
+        type = types.attrsOf (types.submodule ({ config, name, ... }:
+          let profile = config;
+          in {
```
```suggestion
          let

            profile = config;

          in {
```

IldenH commented on modules/programs/firefox.nix:
```diff
@@ -757,6 +765,19 @@ in {
+        assertion = any assertProfile (attrValues cfg.profiles);
+        message = ''
+          'userChrome' won't work with
+          'settings."toolkit.legacyUserProfileCustomizations.stylesheets"' set to false.
```
```suggestion
          {option}`settings."toolkit.legacyUserProfileCustomizations.stylesheets"` set to false.
```

MrQubo replied:
I also added `{option}` to `userChrome`. Not sure if it works? I don't know how to check how it looks like rendered.

MrQubo replied:
At least here, it's not being handled at all. .-.
https://home-manager-options.extranix.com/?query=services.polybar.config&release=master

kira-bruneau replied:
I don't think it makes sense to use `{option}` for assertions. I'm not 100% sure how it's handled for the docs, but assertions don't get rendered like the docs do. They just get dumped out verbatim on the terminal.

IldenH commented on tests/modules/programs/firefox/profile-settings.nix:
```diff
@@ -153,6 +155,11 @@
+
+      profiles.userChrome = {
+        id = 6;
+        userChrome = userChromeContent;
+      };
```
```suggestion
        userChrome = userChromeContent;
        userContent = userChromeContent;
      };
```

IldenH commented on tests/modules/programs/firefox/profile-settings.nix:
```diff
@@ -203,6 +210,14 @@
+      assertFileContains \
+        home-files/.mozilla/firefox/userChrome/chrome/userChrome.css \
+        '${userChromeContent}'
     '';
```
```suggestion

      assertFileContains \
        home-files/.mozilla/firefox/userChrome/chrome/userContent.css \
        '${userChromeContent}'
    '';
```

IldenH commented on modules/programs/firefox.nix:
```diff
@@ -757,6 +765,19 @@ in {
+
+      (let
+        assertProfile = profile:
+          profile.userChrome != ""
```
```suggestion
          profile.userChrome != "" || profile.userContent != ""
```

IldenH commented on modules/programs/firefox.nix:
```diff
@@ -757,6 +765,19 @@ in {
+      in {
+        assertion = any assertProfile (attrValues cfg.profiles);
+        message = ''
+          'userChrome' won't work with
```
```suggestion
          'userChrome' and/or 'userContent' don't work with
```
Unsure on the wording here

MrQubo replied:
I added separate assertions with separate messages. This will be more clear for the users.

IldenH reviewed (changes requested):
Really like this change as it is confusing having the stylesheets not work despite being set.

MrQubo commented:
I hate github, I was replying you, but it was "pending" so I think you didn't see any of that. ;_;

kira-bruneau commented:
It looks like this needs to rebased since: #5128

MrQubo commented:
Rebased, I've made some additional changes, could you review again?

kira-bruneau commented on tests/modules/programs/firefox/profile-settings.nix:
```diff
@@ -154,12 +157,18 @@ in {
       id = 5;
       containers = {
         "shopping" = {
-          id = 6;
```
Wait what was wrong with the containers test? Each container should have its own unique id set, so I think it makes sense to keep one explicitly set in the test.

MrQubo replied:
It was a bit confusing, because it's container id, not profile id. It looks like someone set it to 6 because it's the next number after 5, but these ids are unrelated. I feel like there's no need to set container id here (apart from making this test failproof against changing the default container id), and it's not testing anything.

I could set it explicitly to 1, but using 6 here is a bit confusing, because it looks like it's related to this 5.

kira-bruneau commented on modules/programs/firefox/mkFirefoxModule.nix:
```diff
@@ -801,7 +812,28 @@ in {
+          assertion =
+            all (assertProfile userChromeAttr) (attrValues cfg.profiles);
+          message = ''
+            {option}`${userChromeAttr}` won't work with
+            {option}`settings."toolkit.legacyUserProfileCustomizations.stylesheets"` set to false.
```
I'm not sure if you saw my other comment on this since it was in a resolved thread, but `` {option}`x` `` shouldn't be used in assertions, because they don't get rendered like the docs do.

MrQubo replied:
Ok. I left the backticks characters.

kira-bruneau commented on modules/programs/firefox/mkFirefoxModule.nix:
```diff
@@ -343,410 +343,421 @@ in {
-            '';
-            description = ''
-              Attribute set of ${name} preferences.
+          profile = config;
```
nit: It would cleanup the diff a lot if you just referred to `config` directly rather than giving it an alias of profile, I could see why you'd want to do this to avoid confusion with the top-level `config` though.

rycee replied:
Agreed, a decent middle way may be to introduce the `profile = config` binding inside the `config` attribute. I.e.,

```nix
config = let profile = config; in {
  settings."toolkit.legacyUserProfileCustomizations.stylesheets" =
    mkIf (profile.userChrome != "" || profile.userContent != "") true;
};
```

MrQubo replied:
You're maintainers, so what do you think it's better for maintainability? Using it only in `config` means, that if in the future it won't be possible to refer to `profile` in `settings` in case it's needed e.g. for some `default`.

The bad things are that it will mess up any other PR's that are open right now, and it will also mess up `git blame` (though, there is `git blame -w`, which ignores those indentation changes, not idea whether github uses it).

rycee replied:
I'd say move the let to the config attribute as in my previous comment. We can deal with the (probably unlikely) event that the profile variable needs to be referenced in the options attribute when it happens 🙂

MrQubo replied:
Done

MrQubo commented on modules/programs/firefox/mkFirefoxModule.nix:
```diff
@@ -833,27 +867,27 @@ in {
 
-      "${profilesPath}/${profile.path}/user.js" = mkIf (profile.settings != { }
-        || profile.extraConfig != "" || profile.bookmarks != [ ]) {
+      "${cfg.profilesPath}/${profile.path}/user.js" = mkIf (profile.settings
```
I think it hurts readability how `./format` broke the line before `!=`. :/

MrQubo commented:
I (hopefully) fixed the tests on macos.

MrQubo commented:
Maybe it would be easier to fix the tests in https://github.com/nix-community/home-manager/pull/5698 and keep this PR self-contained.

kira-bruneau commented:
> Maybe it would be easier to fix the tests in https://github.com/nix-community/home-manager/pull/5698 and keep this PR self-contained.

Thanks @MrQubo I'll push the fix to https://github.com/nix-community/home-manager/pull/5698 if it passes here.

kira-bruneau commented:
It looks like it's still failing here

MrQubo commented:
@kira-bruneau I can't understand why it's failing.

MrQubo commented:
Seems like the macos tests where fixed in master.

MrQubo commented:
Now ubuntu tests are broken in master. ;_;

MrQubo commented:
The code moved from `default.nix` to `mkFirefoxModule.nix`, but the changes are still the same.

MrQubo commented:
No idea what is the error now.
It says "NOTE: If your Manual build passes locally and you see this message in CI, you probably need a rebase.", but I already rebased onto master.

nyabinary commented:
> No idea what is the error now. It says "NOTE: If your Manual build passes locally and you see this message in CI, you probably need a rebase", but I already rebased onto master.

Is this working at its current state, or are there still errors?

MrQubo commented:
Tests on this PR need fixing: https://github.com/nix-community/home-manager/pull/6403

MrQubo committed:
firefox: Enable userChrome in about:config

Enable "toolkit.legacyUserProfileCustomizations.stylesheets" in
about:config if userChrome or userContent is not empty.

MrQubo commented:
I decided to not add an assertion after all. The error message about conflicting values of `settings."toolkit.legacyUserProfileCustomizations.stylesheets"` should be enough for the users.
Not having an assertion means users can use `lib.mkForce` to still set it to false if they need to (e.g. maybe someone wants to enable it dynamically with some other method).

MrQubo commented:
I came to fix conflicts and someone merge exactly the same feature. #9532
.-.
